add capability to catch bort errors and return them to an application program

jbathegit opened the pull request:
Fixes #671 

Part of #340
Part of #675

jbathegit committed:
new bort_catcher option, using new borts.c with <setjmp.h>

jbathegit committed:
refactor bort catching logic with full C wrappers to properly utilize <setjmp.h>

jbathegit committed:
add bort catching for several more routines, along with new intest code

jbathegit committed:
fix readns bug which could return a 10-digit jdate even if datelen was never called

jbathegit committed:
fix memory leaks when catching borts inside of certain routines

jbathegit committed:
add recursive label to routines for which bort catching is enabled

jbathegit commented:
@DanielO @jack-woollen

This is an approach that I think will work for the long-standing issue of bort errors within the library which directly terminate the application code.  Instead, with this new approach, an application code can now elect to catch bort errors (see new `src/borts.F90/catch_borts`) and have them returned to the application code, and then the application code can decide for itself what to do.  For backwards-compatibility purposes, the default will still be to directly terminate the application code as is done now, and unless an application code explicitly activates this new feature.  If the feature is activated, it's then incumbent on the application code to do a corresponding call to the new `src/borts.F90/check_for_bort` after every call to a user-callable subroutine, in order to check whether a bort error actually occurred anywhere within the call stack for that routine.

Basically, the way this works is by installing a C-language wrapper around every user-callable subroutine within the library, because C has a standard methodology (using the <setjmp.h> header with the setjmp() and longjmp() functions) which allows for non-local jumps to a previously-set point of the code in case an event like a bort error occurs.  And this non-local jump capability works across multiple levels of the call stack, meaning it will catch and return any bort even if it occurs multiple nested call levels down in a lower-level subroutine, vs. only in the higher-level subroutine that was directly called by the application program and where the jump location was previously set.  The new `test/intest14.F90` code shows how this would work, at least for the subroutines which are included thus far in this PR.

The only caveat is that the call to setjmp() which sets the jump point must still be "active" in the call stack, meaning that whatever C routine you call it from can't have already terminated.  In other words, from inside of Fortran you can't call a C routine to activate setjmp() and then return control back to that Fortran calling routine, because if you do so then that C routine is no longer active and therefore the jump point is lost.  Instead, the Fortran routine has to call the C routine to set the jump point, and then that same C routine has to stay active by directly calling a separate C wrapper (e.g. one of the separate new `src/borts.c/catch_bort_XXXXXX` functions) which in turn will then directly recursively return control back to the appropriate underlying Fortran routine, so that everything is still active in the current call stack in case a subsequent bort error does get triggered.

Unfortunately, standard Fortran has no similar way to do these types of non-local jumps portably to anywhere in a call stack, which is why I had to go with the C-based wrapper approach shown here.  Kudos to @DanielO for pointing out this <setjmp.h> capability in C, as I'd never seen or worked with it before myself.  And so Dan, if we do end up eventually merging this, then I'd like to permanently acknowledge you for that by giving you co-author credit in the library for it, and assuming that's OK with you?

As I noted in #671, I think this approach will work for any application program (whether Fortran, C, or Python) since the logic is now embedded down within the base code of the library itself, as opposed to just in the Python wrapper layer.  All we'll need to do is just develop the appropriate bindings for C and Python, which I think will be easy enough to do, and then we can also develop corresponding test codes which demonstrate how to do that from those languages.

jbathegit commented:
@jack-woollen note also that while testing the new code `test/intest14.F90` for this new capability, I happened to come across a minor bug in readns() where it was always returning jdate as a 10-digit YYYYMMDDHH, and even if the application code had never previously called datelen() to request that.  It was an easy-enough fix that I included as its own separate commit within this PR, for easier future tracking purposes if needed :-)

During testing I also found that, if we ended up activating this feature and catching a bort error, then any memory allocated during a prior call to `src/bufr_c2f_interface.F90/c_f_string` wasn't automatically deallocated, and therefore it would get flagged by the address-sanitizer in the developerCI test.  So I also had to rework a number of the user-callable routines in that source file to use static memory instead, and I then pushed that as yet another separate commit within this PR.

jbathegit commented:
One final note for @jack-woollen - you'll note that I followed the same logic approach that you came up with a few years back for 8-byte integers, using a new local `bort_catch` logical variable that can be toggled on and off within any routine (similar to your previous `im8b` logical variable for 8-byte integers :-), and which ensures that the jump target point will only ever be set within the highest-level subroutine that was directly called by an application program, and not (re)set within any lower-level nested subroutine which may also itself be user-callable within the stack.  For example, if the application code directly calls readmg(), which in turn calls status(), then we always want the jump target point to be at the highest level in readmg(), and not down inside of status() or any other underlying routine.  In other words, we don't successively reset the jump target lower and lower in the call stack, but rather only at the highest level closest to where the user called the library.

I hope that makes sense, especially since otherwise we'd have an even bigger recursion stack to unwind at the end, and besides the whole point here is also to not have to rewrite the entire library to pass return codes all the way back up from a lower level of the call stack and through every intermediate caller in order to ultimately get control all the way back to whatever routine the user directly called! :-)

DanielO commented:
Hi Jeff,
I think it should work - I wonder if it would be more elegant to create a file which contains the externally callable functions and put the setjmp/longjmp in that, and rename the previously exposed internal functions to something else (eg add an "int" suffix or the like).

Obviously it would be a bit tedious to rename all callers internally but it would isolate the kludges better I think.

However I am not really familiar with the library structure so it is hard to be 100% sure it is better.

BTW I notice you have an unrelated change in src/readwritesb.F90 doing a modulus on jdate.

jbathegit commented:
> BTW I notice you have an unrelated change in src/readwritesb.F90 doing a modulus on jdate.

Correct, that's the fix for the minor unrelated bug in readns() that I noted above, where it was always returning jdate as a 10-digit YYYYMMDDHH, and even if the application code had never previously called datelen().  I included that minor fix as a separate commit within this overall PR so that it could be more easily tracked.

jbathegit committed:
updates to README file

jbathegit committed:
add bort catching for ufbint

jbathegit committed:
enable bort catching via character instead of boolean, for future direct callability from C and Python

use local copies of character input triggers, to prevent capit from ever trying to modify character literals

jbathegit committed:
enable bort catching within all Fortran intest and outtest codes

jbathegit commented:
@jack-woollen @AlexanderRichert-NOAA @DanielO

I still have a ways to go with this project, as there are still many user-callable library routines which still need the C-language wrappers to enable bort catching within those routines.  And I also need to develop interfaces and test cases which show how to activate this capability from C and Python application codes, even though I've now laid the groundwork for that by designing the new `catch_borts` and `check_for_bort` routines in such a way that they should be easily callable from either of those languages.

In the meantime, and since we now have all of the Fortran test cases set to automatically test any future bort catching wrappers as soon as they're developed (and since this PR has already reached a critical mass and may soon become completely unwieldy if I continue adding to it :-), I'd like to go ahead and ask for a code review so I can merge what's been developed so far.  I'll then push up a separate future follow-on PR at some point for the remaining updates, once I have a chance to develop and test all of them.

I know this is already a lot to digest, so please take as much time as you need to look this over, and please don't hesitate to ask if you have any questions - thanks!

jbathegit commented on src/standard.F90:
```diff
@@ -38,15 +38,17 @@ subroutine stdmsg(cf)
   implicit none
 
-  character*1, intent(in) :: cf
+  character, intent(in) :: cf
   character*128 bort_str
+  character my_cf
```
Not sure I understand the issue here.  AFAIK w.r.t. standard Fortran, `character` is equivalent to `character*1` or `character(len=1)`, at least insofar as that each of them only ever allows for a single character to be input here.   So I don't agree this needs to be made any more explicit.

jbathegit commented on src/tankrcpt.F90:
```diff
@@ -262,8 +262,9 @@ recursive subroutine strcpt(cf,iyr,imo,idy,ihr,imi)
   implicit none
 
-  character*1, intent(in) :: cf
+  character, intent(in) :: cf
   character*128 bort_str
+  character my_cf
```
As with the related comment for standard.F90, I really don't understand what the issue is here.

I also don't understand what "calling code" it thinks may ever expect it to handle longer strings, b/c it's only a user-callable routine and isn't being directly called internally from anywhere inside of the library(?)

jbathegit commented:
@AlexanderRichert-NOAA @jack-woollen 

Are there any other comments/questions/concerns re: this PR?  If not, then could one of you please approve it, because I'd like to get it merged at some point - thanks! :-)

AlexanderRichert-NOAA commented:
All good here, just pressed the approve button
